### PR TITLE
Shortest path loops

### PR DESCRIPTION
**Pull request by MoKob**
Addressing https://github.com/Project-OSRM/osrm-backend/issues/1863, this pull request fixes a problem were routing actually requires a loop to work correctly.

Opposed to node based routing, starting and ending on a road can require more than just driving along the road itself. In case of one-way streets, we can end up requiring a loop that starts late on the segment and arrives early on the segment. (E.g. how do I get to the beginning of that oneway street).

### CODE REVIEW

**Comment by MoKob** [comment]
General Problem is fixed.
The different search types have to be adjusted to work correctly.

 * Routing Base
- [ ] Search
- [ ] SearchWithCore
- [ ] get_network_distance

 * Alternative Path
- [ ] operator()

 * Many To Many
- [ ] operator()

 * DirectShortestPath
- [x] operator()

 * ShortestPath
- [x] operator()

**Comment by MoKob** [comment]
Turns out that this problem is even more severe then I though. Self loops are a even more general thing. Will have to do a lot more work than I originally thought.

**Comment by MoKob** [comment]
@daniel-j-h could you give this a look. It contains some code to serialize/deserialize flags that I did not need anymore after I had to adjust for all vertices. Maybe that part can still be helpful?

File: features/testbot/via.feature [hunk 1]
```diff
@@ -117,3 +115,67 @@ Feature: Via points
             | 1,2,3     | ab,ab                      | 200m +-1  | head,via,destination                                             |
             | 1,3,2     | ab,ab,bc,cd,da,ab          | 1100m +-1 | head,via,right,right,right,right,destination                     |
             | 3,2,1     | ab,bc,cd,da,ab,ab,bc,cd,da,ab | 1800m     | head,right,right,right,right,via,right,right,right,right,destination |
+
+    Scenario: Via points on ring on the same oneway, forces one of the vertices to be top node
+    # xa it to avoid only having a single ring, which cna trigger edge cases
```
Comment (daniel-j-h) [inline features/testbot/via.feature]: Hmm.. I do not understand that sentence :)

File: include/contractor/contractor.hpp [hunk 2]
```diff
@@ -718,6 +718,50 @@ class Contractor
...
+            // Found a shorter Path -> Update distance
+            else if (to_distance < heap.GetKey(to))
+            {
+                heap.DecreaseKey(to, to_distance);
+                heap.GetData(to).hop = current_hop;
+            }
+        }
+    }
+
+    inline int FindSelfLoop(const NodeID node, int max_distance, ContractorThreadData *const data)
```
Comment (daniel-j-h) [inline include/contractor/contractor.hpp]: Why by pointer and not ref?

File: include/contractor/contractor.hpp [hunk 3]
```diff
@@ -718,6 +718,50 @@ class Contractor
...
+    {
+        ContractorHeap &heap = data->heap;
```
Comment (daniel-j-h) [inline include/contractor/contractor.hpp]: Can be auto

File: include/contractor/contractor.hpp [hunk 4]
```diff
@@ -718,6 +718,50 @@ class Contractor
...
+        heap.Clear();
+        heap.Insert(node, INT_MAX, ContractorHeapData(0, true)); // mark the node itself as a target
```
Comment (daniel-j-h) [inline include/contractor/contractor.hpp]: Hmm int max here depends on the value type used for representation.. this is kind of ugly :/

File: include/contractor/contractor.hpp [hunk 5]
```diff
@@ -718,6 +718,50 @@ class Contractor
...
+        RelaxNode(node, SPECIAL_NODEID, 0, heap);                // add its neighbors to the heap
+        Dijkstra(max_distance, 1, 1000, data, SPECIAL_NODEID);
```
Comment (daniel-j-h) [inline include/contractor/contractor.hpp]: Where do the magic constants for number of targets and especially max nodes come from?
Would be great to at least symbolize them.

File: include/contractor/contractor.hpp [hunk 6]
```diff
@@ -833,6 +852,50 @@ class Contractor
                 continue;
             }
 
+            // Check for Potential Self-Loops, only add if self-loop is the shortest loop
+            for (auto out_edge : contractor_graph->GetAdjacentEdgeRange(node))
+            {
+                const ContractorEdgeData &out_data = contractor_graph->GetEdgeData(out_edge);
```
Comment (daniel-j-h) [inline include/contractor/contractor.hpp]: Could be `const auto&`

File: include/contractor/contractor.hpp [hunk 7]
```diff
@@ -833,6 +852,50 @@ class Contractor
...
+                    if (source == target)
+                    {
+                        // TODO we could prune some of these self-loops if we new about the weight
+                        // along the
+                        // edge represented by `node` itself. If the self-loop is longer than the
+                        // edge,
+                        // the self loop can be pruned as well
+                        // Check for Self Loops
+                        const int this_loop_distance = in_data.distance + out_data.distance;
+                        int self_loop_distance = FindSelfLoop(source, this_loop_distance, data);
```
Comment (daniel-j-h) [inline include/contractor/contractor.hpp]: As far as I can see could also be const

File: include/contractor/contractor.hpp [hunk 8]
```diff
@@ -833,6 +852,50 @@ class Contractor
...
+                            {
+                                inserted_edges.emplace_back(source, target, this_loop_distance,
+                                                            out_data.originalEdges +
+                                                                in_data.originalEdges,
+                                                            node, true, true, false);
+
+                                inserted_edges.emplace_back(target, source, this_loop_distance,
+                                                            out_data.originalEdges +
+                                                                in_data.originalEdges,
+                                                            node, true, false, true);
```
Comment (daniel-j-h) [inline include/contractor/contractor.hpp]: This and above is hard to follow as I have to 1/ look for inserted_edges' value type, 2/ then check which position in the ContractorEdgeData constructor these values correspond to. Hmm probably better to symbolize them, that is create some named booleans and then pass those (especially because only source/target and backward changes).

Would be great to have named parameters á la: fn(forward: true, backward: false) --- oh well :/

File: include/engine/routing_algorithms/alternative_path.hpp [hunk 9]
```diff
@@ -382,10 +382,12 @@ class AlternativeRouting final
         int upper_bound_s_v_path_length = INVALID_EDGE_WEIGHT;
         new_reverse_heap.Insert(via_node, 0, via_node);
         // compute path <s,..,v> by reusing forward search from s
+        const bool constexpr STALLING_ENABLED = true;
+        const bool constexpr DO_NOT_FORCE_LOOPS = false;
```
Comment (daniel-j-h) [inline include/engine/routing_algorithms/alternative_path.hpp]: Nice!

File: include/engine/routing_algorithms/routing_base.hpp [hunk 10]
```diff
@@ -573,29 +634,46 @@ template <class DataFacadeT, class Derived> class BasicRoutingInterface
...
-            RetrievePackedPathFromHeap(forward_heap, reverse_heap, middle, packed_leg);
+            // we need to unpack sub path from core heaps
+            if (facade->IsCoreNode(middle))
+            {
+                std::vector<NodeID> packed_core_leg;
+                RetrievePackedPathFromHeap(forward_core_heap, reverse_core_heap, middle,
+                                           packed_core_leg);
+                BOOST_ASSERT(packed_core_leg.size() > 0);
+                RetrievePackedPathFromSingleHeap(forward_heap, packed_core_leg.front(), packed_leg);
+                std::reverse(packed_leg.begin(), packed_leg.end());
```
Comment (daniel-j-h) [inline include/engine/routing_algorithms/routing_base.hpp]: needs algorithm header; looking at the file, several headers are missing, e.g. vector, iterator, utility, limits, cstddef

File: include/engine/routing_algorithms/routing_base.hpp [hunk 11]
```diff
@@ -573,29 +634,46 @@ template <class DataFacadeT, class Derived> class BasicRoutingInterface
...
     }
 
+    //Requires the heaps for be empty
+    //If heaps should be adjusted to be initialized outside of this function,
+    //the addition of force_loop parameters might be required
     double get_network_distance(SearchEngineData::QueryHeap &forward_heap,
                                 SearchEngineData::QueryHeap &reverse_heap,
                                 const PhantomNode &source_phantom,
-                                const PhantomNode &target_phantom) const
+                                const PhantomNode &target_phantom ) const
```
Comment (daniel-j-h) [inline include/engine/routing_algorithms/routing_base.hpp]: clang-format on all the files

File: include/engine/routing_algorithms/routing_base.hpp [hunk 12]
```diff
@@ -573,29 +634,46 @@ template <class DataFacadeT, class Derived> class BasicRoutingInterface
...
     {
+        BOOST_ASSERT( forward_heap.Size() == 0 );
+        BOOST_ASSERT( reverse_heap.Size() == 0 );
```
Comment (daniel-j-h) [inline include/engine/routing_algorithms/routing_base.hpp]: .Empty

File: include/engine/routing_algorithms/shortest_path.hpp [hunk 13]
```diff
@@ -255,7 +169,10 @@ class ShortestPathRouting final
             BOOST_ASSERT(forward_heap.Size() > 0);
             BOOST_ASSERT(reverse_heap.Size() > 0);
             super::Search(forward_heap, reverse_heap, new_total_distance_to_reverse,
-                          leg_packed_path_reverse);
+                          leg_packed_path_reverse, false,
+                          source_phantom.reverse_node_id == target_phantom.reverse_node_id &&
+                              source_phantom.GetReverseWeightPlusOffset() >
+                                  target_phantom.GetReverseWeightPlusOffset());
```
Comment (daniel-j-h) [inline include/engine/routing_algorithms/shortest_path.hpp]: Wait.. this and above is your forceLoop lambda body, isn't it?

File: include/extractor/extractor.hpp [hunk 14]
```diff
@@ -11,14 +11,15 @@ namespace osrm
...
-class extractor
+class Extractor
 {
   public:
-    extractor(ExtractorConfig extractor_config) : config(std::move(extractor_config)) {}
+    Extractor(ExtractorConfig extractor_config) : config(std::move(extractor_config)) {}
     int run();
 
-  private:
+  protected:
```
Comment (daniel-j-h) [inline include/extractor/extractor.hpp]: Why protected?

File: include/util/io.hpp [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+             ++chunk_bit, ++bit_nr)
+            chunk_bitset[chunk_bit] = flags[bit_nr];
+
+        chunk = chunk_bitset.to_ulong();
+        ++chunk_count;
+        flag_stream.write(reinterpret_cast<const char *>(&chunk), sizeof(chunk));
+    }
+    util::SimpleLogger().Write() << "Wrote " << number_of_bits << " bits in " << chunk_count
+                                 << " chunks (Flags).";
+    return !flag_stream.fail();
```
Comment (daniel-j-h) [inline include/util/io.hpp]: Hmm I would be careful with explicitely checking states here; if you just want to verify the I/O operation went fine and the stream is okay, then force the stream to a boolean context, such as `if(stream) ok();`

File: include/util/io.hpp [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+    for (std::size_t chunk_id = 0; chunk_id < chunks; ++chunk_id)
+    {
+        flag_stream.read(reinterpret_cast<char *>(&chunk), sizeof(chunk));
+        std::bitset<32> chunk_bits(chunk);
+        for (std::size_t bit = 0; bit < 32 && bit_position < number_of_bits; ++bit, ++bit_position)
+            flags[bit_position] = chunk_bits[bit];
+    }
+    util::SimpleLogger().Write() << "Read " << number_of_bits << " bits in " << chunks
+                                 << " Chunks from disk.";
+    return !flag_stream.fail();
```
Comment (daniel-j-h) [inline include/util/io.hpp]: Same here

File: src/contractor/processing_chain.cpp [hunk 17]
```diff
@@ -143,10 +147,12 @@ std::size_t Prepare::LoadEdgeExpandedGraph(
...
-    size_t number_of_edges = 0;
-    size_t max_edge_id = SPECIAL_EDGEID;
-    input_stream.read((char *)&number_of_edges, sizeof(size_t));
-    input_stream.read((char *)&max_edge_id, sizeof(size_t));
+    // TODO std::size_t can vary on systems. Our files are not transferable, but we might want to
+    // consider using a fixed size type for I/O
+    std::size_t number_of_edges = 0;
+    std::size_t max_edge_id = SPECIAL_EDGEID;
+    input_stream.read((char *)&number_of_edges, sizeof(std::size_t));
+    input_stream.read((char *)&max_edge_id, sizeof(std::size_t));
```
Comment (daniel-j-h) [inline src/contractor/processing_chain.cpp]: Looking at `ag -Q 'sizeof(unsigned'` we're writing size prefixes as unsigneds most if not all of the time

File: include/contractor/contractor.hpp (hunk 2 above)
Comment (MoKob) [inline include/contractor/contractor.hpp]: mostly due to consistency. It was used like this everywhere else in the contractor. That thing really requires a refactor.

**Comment by MoKob** [comment]
While this works, it significantly increases preprocessing time (Germany 662 -> 781 seconds).

A possible way to improve this would be to remember self-loop prune distances based on the length of the ways represented in the edge based nodes, and INFTY for one-ways. This requires updating these values in case of traffic updates as well. @TheMarex @danpat @daniel-j-h opinions on this?

**Comment by MoKob** [comment]
Got the increase down to 697.7 from 662 (previously 781) on the German road network.
After review, this should solve the loop problems.

**Comment by MoKob** [comment]
With the new values, sadly the performance is back down to 746.107 seconds.

**Comment by MoKob** [comment]
Last commit got it down to 700s again. Yay me.

File: include/contractor/contractor.hpp [hunk 18]
```diff
@@ -846,22 +869,65 @@ class Contractor
...
+                            stats->original_edges_added_count +=
+                                2 * (out_data.originalEdges + in_data.originalEdges);
+                        }
+                        else
+                        {
+                            // CAREFUL: This only works due to the independent node-setting. This
+                            // guarantees that source is not connected to another node that is
+                            // contracted
+                            node_weights[source] = loop_weight + 1;
+                            node_weights[source] = loop_weight; // make sure to prune better
```
Comment (TheMarex) [inline include/contractor/contractor.hpp]: This looks like a typo, ` node_weights[source] ` gets assigned twice.

File: include/contractor/contractor.hpp [hunk 19]
```diff
@@ -846,22 +869,65 @@ class Contractor
...
+
+                            inserted_edges.emplace_back(
+                                target, source, loop_weight,
+                                out_data.originalEdges + in_data.originalEdges, node, SHORTCUT_ARC,
+                                FORWARD_DIRECTION_DISABLED, REVERSE_DIRECTION_ENABLED);
+                        }
+                    }
+                    continue;
+                }
                 const int path_distance = in_data.distance + out_data.distance;
```
Comment (TheMarex) [inline include/contractor/contractor.hpp]: `loop_weight` and `path_distance` seem to always have the same value, but are declared in different scopes. Move `path_distance` up.

File: include/engine/routing_algorithms/many_to_many.hpp [hunk 20]
```diff
@@ -143,10 +143,30 @@ class ManyToManyRouting final
                     (*result_table)[source_id * number_of_targets + target_id];
                 // check if new distance is better
                 const EdgeWeight new_distance = source_distance + target_distance;
-                if (new_distance >= 0 && new_distance < current_distance)
+                if (new_distance < 0)
                 {
-                    (*result_table)[source_id * number_of_targets + target_id] =
-                        (source_distance + target_distance);
+                    for (auto edge : super::facade->GetAdjacentEdgeRange(node))
```
Comment (TheMarex) [inline include/engine/routing_algorithms/many_to_many.hpp]: I feel this should be a utility function in `routing_base.hpp`. Something like:

`UpdateLoopWeight(NodeId node, EdgeWeight& loop_weight)`

Also can't we guarantee that each node has at most one relevant loop? In that case we can optimize this.

Comment (MoKob) [inline include/engine/routing_algorithms/many_to_many.hpp]: We could increase the memory overhead to have a lookup. Since this should only happen at the start of a segment, I guess the overhead to compute it, is worth it to save the additional entry for every node.

Moved the lookup to Routing base.
